In [0]:
#all spark imports
from pyspark.sql import SparkSession
from pyspark.sql.types import *
from pyspark.sql.functions import *

In [0]:
#instantiate the spark session
spark = SparkSession.builder.appName("AssignmentTwo").getOrCreate()

1. Count the odd and even numbers usingfile‘integer.txt’download it from the Quercus. Show your code and output.

In [0]:
integerRDD = spark.sparkContext.textFile("/FileStore/tables/integer.txt", 4)
arrayRDD = integerRDD.map(lambda x: ( int(x)))
evenCount=arrayRDD.filter(lambda x: x%2 == 0).count()
print("EvenCount: %3d"% (evenCount))
oddcount=arrayRDD.filter(lambda x: x%2 != 0).count()
print("OddCount: %3d"% (oddcount))

EvenCount: 514
OddCount: 496


2 Calculate the salary sum per department using file ‘salary.txt’ download it from the Quercus. Show department name and salary sum. Show your code and output.

In [0]:
salaryRDD = spark.sparkContext.textFile("/FileStore/tables/salary.txt", 4)
arrayRDD = salaryRDD.map(lambda x: x.split(" "))
kvRDD = arrayRDD.map(lambda x: (x[0], float(x[1])))
sumRDD = kvRDD.reduceByKey(lambda x,y: x+y)
sumRDD.collect()

Out[4]: [('Developer', 3221394.0),
 ('Sales', 3488491.0),
 ('Research', 3328284.0),
 ('QA', 3360624.0),
 ('Marketing', 3158450.0)]

3 Implement MapReduce using Pyspark on file ‘shakespeare.txt’ download it from the Quercus. Show how many times these particular words appear in the document: Shakespeare, When, Lord, Library, GUTENBERG, WILLIAM, COLLEGE and WORLD. (Count exact words only)

In [0]:
import re 
shakespeareRDD = spark.sparkContext.textFile("/FileStore/tables/shakespeare_1-1.txt", 4)
counts=shakespeareRDD.flatMap(lambda line: line.split(" ")) \
                  .filter(lambda word: word == "When" or word=="Lord" or word == "Library" or word == "GUTENBERG" or word == "WILLIAM" or word == "COLLEGE" or word == "WORLD") \
                  .map(lambda word: (word, 1)) \
                  .reduceByKey(lambda a, b: a +b)

for element in counts.collect():
    print(element)

    


    



('WILLIAM', 115)
('WORLD', 98)
('When', 393)
('GUTENBERG', 99)
('COLLEGE', 98)
('Lord', 341)
('Library', 2)


4 Calculate top15 and bottom15 words using file‘shakespeare.txt’ download it from the Quercus. Show 15 words with most count and 15 words with least count. You can limit by 15 in ascending and descending order of count. Show your code and output.

In [0]:
shakespeareRDD = spark.sparkContext.textFile("/FileStore/tables/shakespeare_1-1.txt", 4)
counts=shakespeareRDD.flatMap(lambda line: line.split(" ")) \
                  .map(lambda word: (word, 1)) \
                  .reduceByKey(lambda a, b: a +b)\
                  .top(15,  key=lambda word: word[1])


for element in counts:
    print(element)

('', 231583)
('the', 11397)
('and', 8777)
('I', 8556)
('of', 7873)
('to', 7421)
('a', 5672)
('my', 4913)
('in', 4600)
('you', 4060)
('And', 3547)
('that', 3522)
('is', 3481)
('his', 3226)
('with', 3175)


In [0]:
shakespeareRDD = spark.sparkContext.textFile("/FileStore/tables/shakespeare_1-1.txt", 4)
counts=shakespeareRDD.flatMap(lambda line: line.split(" ")) \
                  .map(lambda word: (word, 1)) \
                  .reduceByKey(lambda a, b: a +b)\
                  .takeOrdered(15,  key=lambda word: word[1])


for element in counts:
    print(element)

('whatsoever.', 1)
('re-use', 1)
('eBook,', 1)
('file.', 1)
('Author:', 1)
('Posting', 1)
('1,', 1)
('Language:', 1)
('Character', 1)
('cooperation', 1)
('*This', 1)
('read!*', 1)
('SHAREWARE', 1)
('PUBLIC', 1)
('DOMAIN.', 1)


In [0]:
from pyspark.ml.feature import Imputer, StringIndexer, VectorAssembler
from pyspark.ml.linalg import SparseVector, DenseVector
from pyspark.ml import Pipeline
from pyspark.ml.tuning import CrossValidator, ParamGridBuilder, TrainValidationSplit
from pyspark.ml.recommendation import ALS
from pyspark.ml.evaluation import RegressionEvaluator


1. Describe your data. Calculate top 12 movies with highest ratings and top 12 users who provided highest ratings. Show your code and output.


The data has 1501 rows and three columns. The columns in the data are movieId, ratings user give for the movie and userId
Data type of Columns are: 'movieId', 'int', 'rating', 'int', 'userId', 'int'.

In [0]:
path = "/FileStore/tables/movies.csv"
 
df = spark.read \
  .format("csv") \
  .option("inferSchema", True) \
  .option("header", True) \
  .option("path", path) \
  .load()

display(df)
# number of rows from the Dataframe
row = df.count()
# number of columns from the Dataframe
col = len(df.columns)

# data type of columns
datatype=df.dtypes
# printing
print(f'Dimension of the Dataframe is: {(row,col)}')
print(f'Number of Rows are: {row}')
print(f'Number of Columns are: {col}')
print(f'Data type of Columns are: {datatype}')


Dimension of the Dataframe is: (1501, 3)
Number of Rows are: 1501
Number of Columns are: 3
Data type of Columns are: [('movieId', 'int'), ('rating', 'int'), ('userId', 'int')]


movieId rating userId 2 3 0 3 1 0 5 2 0 9 4 0 11 1 0 12 2 0 15 1 0 17 1 0 19 1 0 21 1 0 23 1 0 26 3 0 27 1 0 28 1 0 29 1 0 30 1 0 31 1 0 34 1 0 37 1 0 41 2 0 44 1 0 45 2 0 46 1 0 47 1 0 48 1 0 50 1 0 51 1 0 54 1 0 55 1 0 59 2 0 61 2 0 64 1 0 67 1 0 68 1 0 69 1 0 71 1 0 72 1 0 77 2 0 79 1 0 83 1 0 87 1 0 89 2 0 91 3 0 92 4 0 94 1 0 95 2 0 96 1 0 98 1 0 99 1 0 2 2 1 3 1 1 4 2 1 6 1 1 9 3 1 12 1 1 13 1 1 14 1 1 16 1 1 19 1 1 21 3 1 27 1 1 28 3 1 33 1 1 36 2 1 37 1 1 40 1 1 41 2 1 43 1 1 44 1 1 47 1 1 50 1 1 54 1 1 56 2 1 57 1 1 58 1 1 60 1 1 62 4 1 63 1 1 67 1 1 68 4 1 70 2 1 72 1 1 73 1 1 74 2 1 76 1 1 77 3 1 78 1 1 81 1 1 82 1 1 85 3 1 86 2 1 88 2 1 91 1 1 92 2 1 93 1 1 94 2 1 96 1 1 97 1 1 4 3 2 6 1 2 8 5 2 9 1 2 10 1 2 12 3 2 13 1 2 15 2 2 18 2 2 19 4 2 22 1 2 26 1 2 28 1 2 34 4 2 35 1 2 37 5 2 38 1 2 39 5 2 40 4 2 47 1 2 50 1 2 52 2 2 54 1 2 55 1 2 57 2 2 58 2 2 59 1 2 61 1 2 62 1 2 64 1 2 65 1 2 66 3 2 68 1 2 71 3 2 76 1 2 77 1 2 78 1 2 80 1 2 83 5 2 85 1 2 87 2 2 88 1 2 89 4 2 90 1 2 92 4 2 93 5 2 0 1 3 1 1 3 2 1 3 7 3 3 8 3 3 9 1 3 14 1 3 15 1 3 16 1 3 18 4 3 19 1 3 24 3 3 26 1 3 29 3 3 33 1 3 34 3 3 35 1 3 36 3 3 37 1 3 38 2 3 43 1 3 44 1 3 46 1 3 47 1 3 51 5 3 52 3 3 56 1 3 58 1 3 60 3 3 62 1 3 65 2 3 66 1 3 67 1 3 68 2 3 70 1 3 72 2 3 76 3 3 79 3 3 80 4 3 81 1 3 83 1 3 84 1 3 86 1 3 87 2 3 88 4 3 89 1 3 91 1 3 94 3 3 1 1 4 6 1 4 8 1 4 9 1 4 10 1 4 11 1 4 12 1 4 13 1 4 14 2 4 15 1 4 17 1 4 20 1 4 22 1 4 23 1 4 24 1 4 29 4 4 30 1 4 31 1 4 34 1 4 35 1 4 36 1 4 39 2 4 40 3 4 41 4 4 43 2 4 44 1 4 45 1 4 46 1 4 47 1 4 49 2 4 50 1 4 51 1 4 52 4 4 54 1 4 55 1 4 60 3 4 61 1 4 62 4 4 63 3 4 65 1 4 67 2 4 69 1 4 70 4 4 71 1 4 73 1 4 78 1 4 84 1 4 85 1 4 87 3 4 88 3 4 89 2 4 96 1 4 97 1 4 98 1 4 99 1 4 0 1 5 1 1 5 4 1 5 5 1 5 8 1 5 9 3 5 10 2 5 13 3 5 15 1 5 19 1 5 20 3 5 21 2 5 23 3 5 27 1 5 28 1 5 29 1 5 31 1 5 36 3 5 38 2 5 39 1 5 42 1 5 48 3 5 49 4 5 50 3 5 51 1 5 52 1 5 54 1 5 55 5 5 56 3 5 58 1 5 60 1 5 61 1 5 64 3 5 65 2 5 68 4 5 70 1 5 71 1 5 72 1 5 74 1 5 79 1 5 81 2 5 84 1 5 85 1 5 86 1 5 88 1 5 90 4 5 91 2 5 95 2 5 99 1 5 0 1 6 1 1 6 2 3 6 5 1 6 6 1 6 9 1 6 10 1 6 15 2 6 16 2 6 17 1 6 18 1 6 20 1 6 21 1 6 22 1 6 24 1 6 25 5 6 26 1 6 28 1 6 30 1 6 33 1 6 38 1 6 39 1 6 43 4 6 44 1 6 45 1 6 48 1 6 49 1 6 50 1 6 53 1 6 54 1 6 55 1 6 56 1 6 58 4 6 59 1 6 60 1 6 61 3 6 63 3 6 66 1 6 67 3 6 68 1 6 69 1 6 71 2 6 73 1 6 75 1 6 77 1 6 79 1 6 81 1 6 84 1 6 85 3 6 86 1 6 87 1 6 88 1 6 89 1 6 91 2 6 94 1 6 95 2 6 96 1 6 1 1 7 2 2 7 3 1 7 4 1 7 7 1 7 10 1 7 11 2 7 14 2 7 15 1 7 16 1 7 18 1 7 21 1 7 22 1 7 23 1 7 25 5 7 26 1 7 29 4 7 30 1 7 31 3 7 32 1 7 33 1 7 35 1 7 37 2 7 39 3 7 40 2 7 42 2 7 44 1 7 45 2 7 47 4 7 48 1 7 49 1 7 53 1 7 54 1 7 55 1 7 56 1 7 59 1 7 61 2 7 62 3 7 63 2 7 66 1 7 67 3 7 74 1 7 75 1 7 76 3 7 77 1 7 81 1 7 82 1 7 84 2 7 85 4 7 86 1 7 92 2 7 96 1 7 97 1 7 98 1 7 0 1 8 2 4 8 3 2 8 4 2 8 5 1 8 7 1 8 9 1 8 11 1 8 15 1 8 18 1 8 19 1 8 21 1 8 29 5 8 31 3 8 33 1 8 35 1 8 36 1 8 40 2 8 44 1 8 45 1 8 50 1 8 51 1 8 52 5 8 53 5 8 54 1 8 55 1 8 56 1 8 58 4 8 60 3 8 62 4 8 64 1 8 67 3 8 69 1 8 71 1 8 72 3 8 77 3 8 78 1 8 79 1 8 83 1 8 85 5 8 86 1 8 88 1 8 90 1 8 92 2 8 95 4 8 96 3 8 97 1 8 98 1 8 99 1 8 2 3 9 3 1 9 4 1 9 5 1 9 6 1 9 7 5 9 9 1 9 12 1 9 14 3 9 15 1 9 19 1 9 21 1 9 22 1 9 24 1 9 25 1 9 26 1 9 30 3 9 32 4 9 35 2 9 36 2 9 37 2 9 38 1 9 39 1 9 43 3 9 49 5 9 50 3 9 53 1 9 54 1 9 58 1 9 59 1 9 60 1 9 61 1 9 63 3 9 64 3 9 68 1 9 69 1 9 70 3 9 71 1 9 73 2 9 75 1 9 77 2 9 81 2 9 82 1 9 83 1 9 84 1 9 86 1 9 87 4 9 88 1 9 90 3 9 94 2 9 95 3 9 97 2 9 98 1 9 0 3 10 2 4 10 4 3 10 7 1 10 8 1 10 10 1 10 13 2 10 14 1 10 16 2 10 17 1 10 18 1 10 21 1 10 22 1 10 24 1 10 25 3 10 28 1 10 35 1 10 36 1 10 37 1 10 38 1 10 39 1 10 40 4 10 41 2 10 42 3 10 43 1 10 49 3 10 50 1 10 51 1 10 52 1 10 55 2 10 56 1 10 58 1 10 63 1 10 66 1 10 67 2 10 68 1 10 75 1 10 77 1 10 79 1 10 86 1 10 89 3 10 90 1 10 97 1 10 98 1 10 0 1 11 6 2 11 9 1 11 10 1 11 11 1 11 12 1 11 13 4 11 16 1 11 18 5 11 19 4 11 20 1 11 21 1 11 22 1 11 23 5 11 25 1 11 27 5 

In [0]:
df1=df.groupBy("movieId").sum('rating')

display(df1.sort("sum(rating)").tail(12))


df2=df.groupBy("userId").sum("rating")
display(df2.sort("sum(rating)").tail(12))

movieId sum(rating) 62 36 18 36 88 37 23 37 49 39 51 40 2 40 22 41 68 41 90 45 94 47 29 48

userId sum(rating) 18 90 28 91 8 93 9 95 2 95 24 98 12 102 14 102 26 108 23 111 22 121 11 128

2 Split dataset into train and test. Try 2 different combinations for e.g. (60/40, 70/30, 75/25 and 80/20). (Train your model and use collaborative filtering approach on 70 percent of your data and test with the other 30 percent and so on). Show your code and output.

In [0]:

# split data into training and test datasets 
train_df1, test_df1= df.randomSplit([0.7, 0.3], 1234)
train_df2, test_df2=df.randomSplit([0.8, 0.2], 1234)
# Build the recommendation model using ALS on the training data
# Note we set cold start strategy to 'drop' to ensure we don't get NaN evaluation metrics
als1 = ALS(maxIter=5, regParam=0.01,userCol="userId", itemCol="movieId", ratingCol="rating", coldStartStrategy="drop")

model1 = als1.fit(train_df1)
model2=als1.fit(train_df2)





3  Explain MSE, RMSE and MAE. Compare and evaluate both of your models with evaluation metrics (RMSE or MAE), show your code and print your results. Describe which one works better and why?

All  three parameteres help evaluate  model accuracy

MSE (Mean Square Error): square difference between the orginal value and predicted value. The lesser the value of MSE the more close the predicted value is to the orginal value.
RMSE (Root Mean Square Error) which is the square root of MSE. the lesser the result of RMSE the more accurate the model is. The result of RMSE states that the orginal value is +/- RMSE  of the predicted value
MAE(Mean Absolute Error): It is the average difference between the orginal and pridicted value.

In [0]:
# Evaluate the model by computing the RMSE on the test data for model 0ne
predictions1 = model1.transform(test_df1)

evaluator1 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse1 = evaluator1.evaluate(predictions1)
print("Root-mean-square error for model one (0.7,0.3) = " + str(rmse1))



# Evaluate the model by computing the RMSE on the test data for model Two
predictions2 = model2.transform(test_df2)

evaluator2 = RegressionEvaluator(metricName="rmse", labelCol="rating",
                                predictionCol="prediction")
rmse2 = evaluator2.evaluate(predictions2)
print("Root-mean-square error for model two (0.8,0.2) = " + str(rmse2))

Root-mean-square error for model one (0.7,0.3) = 1.9847788424275474
Root-mean-square error for model two (0.8,0.2) = 1.6218664580955613


RMSE value for model one which is the data is split 70% training and 30% test data is approximatly 1.985 as shown above in the output. On the other hand RMSE for model two which is 80% training and 20% test data resulted 1.62 RMSE. The result suggest that the higher the proportion of the training dataset the more accurate the model is.

4 Now tune the parameters of your algorithm to get the best set of parameters. Explain different parameters of the algorithm which you have used for tuning your algorithm. Evaluate all your models again. Show your code with best values and output.

In [0]:
train_df, test_df= df.randomSplit([0.7, 0.3], 1234)
als = ALS(userCol="userId", itemCol="movieId", ratingCol="rating", coldStartStrategy="drop")

#Tune model using ParamGridBuilder 
paramGrid = ParamGridBuilder()\
    .addGrid(als.regParam, [0.1, 0.01]) \
    .addGrid(als.rank, [10, 50])\
    .addGrid(als.maxIter, [1,3])\
    .build()
evaluators = RegressionEvaluator(metricName="rmse", labelCol="rating",predictionCol="prediction")
# build cross validation
cv = CrossValidator(estimator =als, estimatorParamMaps = paramGrid,evaluator =evaluators, numFolds=3)
#Fit model to training data
model = cv.fit(train_df)
# Evaluate the model by computing the RMSE on the test data for the model
predictions = model.transform(test_df)
rmse = evaluators.evaluate(predictions)
#Extract best model from the cv model above
bestModel = model.bestModel
print("**Best Model**")
# Print "Rank"
print("  Rank:", bestModel._java_obj.parent().getRank())
# Print "MaxIter"
print("  MaxIter:", bestModel._java_obj.parent().getMaxIter())
# Print "RegParam"
print("  RegParam:", bestModel._java_obj.parent().getRegParam())
print("Root-mean-square error for the  model  = " + str(rmse))

/databricks/spark/python/pyspark/ml/util.py:839: UserWarning: Cannot find mlflow module. To enable MLflow logging, install mlflow from PyPI.
  warnings.warn(_MLflowInstrumentation._NO_MLFLOW_WARNING)
**Best Model**
  Rank: 50
  MaxIter: 3
  RegParam: 0.1
Root-mean-square error for the  model  = 1.0505386361339517


The different parameters used inorder to tune our model are : 
    regParam, [0.1, 0.01]) : which is a regularization. parametere default is 1.0 I have used set of 0.1 and 0.01 
    rank, [10, 50]: is the number of latent factor in the model default is 10. I have used set of 10,50 in the gridBuilder
    maxIter, [1,3]: is number of times your model iterates default to 10 and in the gridBuilder we have used set of 1 and 3

4  Calculate top 12 movies recommendations for userid 1 and userid 12. Show your code and output.

In [0]:
# a list of movies we are thinking to offer
user1_suggest = test_df.filter(train_df['userId']==1).select(['movieId', 'userId'])
user12_suggest = test_df.filter(train_df['userId']==12).select(['movieId', 'userId'])
# Top 12 movies recommndations for user one 
user1_recommendations= model.transform(user1_suggest)
display(user1_recommendations.orderBy('prediction').tail(12))


# Top 12 movies recommndations for user 12
user12_recommendations= model.transform(user12_suggest)
display(user12_recommendations.orderBy('prediction', decending=False).limit(12))

movieId userId prediction 6 1 1.2027411460876465 13 1 1.291556477546692 40 1 1.2950832843780518 93 1 1.3220633268356323 73 1 1.3668843507766724 9 1 1.4442086219787598 67 1 1.6449387073516846 56 1 1.7549599409103394 74 1 1.7646316289901733 63 1 1.8299444913864136 36 1 1.8922770023345947 77 1 1.9474213123321533

movieId userId prediction 31 12 0.65923315 78 12 0.69465554 14 12 0.76410866 86 12 0.7757331 21 12 0.8061414 15 12 0.84743696 4 12 1.0146977 24 12 1.1157546 82 12 1.2705686 51 12 1.2999612 56 12 1.5756509 22 12 1.8342023